### PR TITLE
Refactor Signing Rules

### PR DESCRIPTION
**Pull request by cfrantz**
Refactor the offline and online signing rules to use a common set of helper functions and use a common flow.

1. Refactor the offline signing rules into a set of common functions that can be used for both offline and online signing flows.
2. Move the `sign_bin` rule to signing.bzl.
3. Refactor the rule to perform signing using the same flow as the offline process:
   - Generate pre-signing artifacts.
   - Sign the artifacts, yielding detached signatures.
   - Attach the detached signatures to the pre-signing binaries.

This refactoring of the `sign_bin` rule is to prepare for signing test binaries with keys held on personal tokens (e.g. Nitrokeys).

### CODE REVIEW

File: signing/examples/BUILD [hunk 1]
```diff
@@ -17,7 +17,9 @@ offline_presigning_artifacts(
     # key (e.g. //sw/device/silicon_creator/rom/keys/real/rsa:earlgrey_a0_test_0).
     # This is left as a fake key so that the presigning artifacts will be
     # appropriate for the `fake_sign` rule below.
-    rsa_key = "//sw/device/silicon_creator/rom/keys/fake/rsa:test_private_key_0",
+    rsa_key = {
+        "//sw/device/silicon_creator/rom/keys/fake/rsa:test_private_key_0": "fake_test_key_0",
```
Comment (moidx) [inline signing/examples/BUILD:21]: qq to make sure I understand: 

1. Is the value only used to generate the produced filename suffix? 
2. Should we add an example to drive sining with a real test key? Assuming the target just changes here to point to a public key.

**Review by moidx — APPROVED** [review]

File: signing/examples/BUILD (hunk 1 above)
Comment (cfrantz) [reply signing/examples/BUILD:21]: Currently, this value is used to derive the filename suffix.  When I add support for signing with a token, the label will be the keyset (which will contain the public key information) and the value will be the PKCS#11 label of the signing key.

I'm not completely satisfied that this field can have different inputs depending on the key type (local fake private key vs. on-token keys), but this allows us to maintain compatibility with the existing signing methodology test programs.

File: rules/signing.bzl [hunk 2]
```diff
@@ -2,76 +2,234 @@
...
+def key_ext(rsa, spx):
+    if spx:
+        return ".{}.{}".format(rsa.name, spx.name)
+    else:
+        return ".{}".format(rsa.name)
+
+def _presigning_artifacts(ctx, opentitantool, src, manifest, rsa_key, spx_key, basename = None, keyname_in_filenames = False):
+    """Create the pre-signing artifacts for a given input binary.
+
+    Applies he manifest and public components of the keys.  Creates the
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply rules/signing.bzl]: Fixed.

File: signing/examples/BUILD (hunk 1 above)
Comment (moidx) [reply signing/examples/BUILD:21]: ack. I missed the dual use aspect of the dict value. We should capture the conventions in a follow up PR as part of the signing documentation.

**Review by timothytrippel — APPROVED** [review]

File: rules/signing.bzl [hunk 3]
```diff
@@ -3,75 +3,105 @@
...
+    digests/messages required for signing.
+
+    Args:
+        opentitantool: file; The opentitantool binary.
+        src: file; The source binary
+        manifest: file; The manifest file.
+        rsa_key_file: file; The RSA public key.
+        spx_key_file: file; The SPX+ public key.
+    Returns:
+        (file, file, file): The pre-signing binary, the digest and spx message.
```
Comment (alphan) [inline rules/signing.bzl:49]: nit: a mismatch would probably be very easy to catch but I'd prefer a struct or dict return type for clarity.

File: sw/host/opentitantool/src/command/image.rs [hunk 4]
```diff
@@ -223,7 +231,9 @@ impl CommandDispatch for ManifestUpdateCommand {
         }
 
         // Update `length` to be able to generate signatures.
-        image.update_length()?;
```
Comment (alphan) [inline sw/host/opentitantool/src/command/image.rs:226]: does this conflict with @jon-flatley's recent PR?

File: signing/rom_tests/signatures/empty_test_sigverify_spx_fpga_cw310.rsa-sig
Comment (alphan) [inline signing/rom_tests/signatures/empty_test_sigverify_spx_fpga_cw310.rsa-sig:1]: nit: `rsa_sig` instead of `rsa-sig` for the sake of consistency?

**Review by alphan — APPROVED** [review]
Thanks @cfrantz !

File: rules/signing.bzl (hunk 3 above)
Comment (cfrantz) [reply rules/signing.bzl:49]: fixed.

File: sw/host/opentitantool/src/command/image.rs (hunk 4 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/image.rs:226]: Probably.  I'll wait for Jon's PR to submit and then update mine accordingly.

File: signing/rom_tests/signatures/empty_test_sigverify_spx_fpga_cw310.rsa-sig
Comment (cfrantz) [reply signing/rom_tests/signatures/empty_test_sigverify_spx_fpga_cw310.rsa-sig:1]: done

File: sw/host/opentitantool/src/command/image.rs (hunk 4 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/image.rs]: I've rebased now that #18893 is in and ran a couple of tests locally.  Re-pushed to run everything through CI.

**Pull request merged**